prometheusbp/high_watermark.go

```diff
@@ -0,0 +1,111 @@
+package prometheusbp
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// HighWatermarkValue implements an int64 gauge with high watermark value.
```

Comment:
I think this exposes too much of this API surface and makes it too hard to use.

I would use prometheus.Histogram as a guide here.  It's exposed as a single API but when collected it produces more than one metric with a fixed naming scheme.

Expose:
* `func NewHighWaterMarkGauge(HighWatermarkGaugeArgs) *HighWatermarkGauge`
* `func NewHighWaterMarkGaugeVec(HighWatermarkGaugeArgs, labels) *HighWatermarkGaugeVec`
* `type HighWatermarkGaugeVec`
  * With / WithLabels
* `type HighWatermarkGauge`
  * Add / Sub / Set
  * Inc / Dec
  * Collect

Reply:
I would strongly prefer to separate the API for the value and api for the actual prometheus metrics (the 2 gauges). There's no guarantee that we would always define the same set of labels for the 2 gauges, so an API like `prometheus.Histogram` can be very brittle.

As for `Add`/`Sub`, I just don't really see a use case for them yet. If we do have the use case for them later we can add them when then.

Reply:
> There's no guarantee that we would always define the same set of labels for the 2 gauges

I would approach this from the _other_ side.  Why would we _not_ want them to have the same labels?  Unlike a histogram where you may want a counter with additional labels, we're not talking about exploding cardinality here, and it would be pretty confusing to have different labels for the instant and high watermark metrics.